// wa-map-optimizer-vite/dist/index.js
import fs from "fs";
import path from "path";
import { optimize } from "wa-map-optimizer";
import crypto from "crypto";
import { ITiledMap } from "@workadventure/tiled-map-type-guard";
export { LogLevel } from "wa-map-optimizer/dist/guards/libGuards.js";
export function getMaps(mapDirectory = ".") {
    let mapFiles = new Map();
    for (const file of fs.readdirSync(mapDirectory)) {
        const fullPath = mapDirectory + "/" + file;
        if (mapDirectory && fs.lstatSync(fullPath).isDirectory() && file !== "dist" && file !== "node_modules") {
            mapFiles = new Map([...mapFiles, ...getMaps(fullPath)]);
        }
        else {
            const map = isMapFile(fullPath);
            if (!map) {
                continue;
            }
            mapFiles.set(fullPath, map);
        }
    }
    return mapFiles;
}
function isMapFile(filePath) {
    if (!filePath.endsWith(".tmj")) {
        return undefined;
    }
    let object;
    try {
        object = JSON.parse(fs.readFileSync(filePath).toString());
    }
    catch (error) {
        throw new Error(`Error on ${filePath} map file`, { cause: error });
    }
    const mapFile = ITiledMap.safeParse(object);
    if (!mapFile.success) {
        console.error(`${filePath} is not a compatible map file, the file will be skip`);
        console.error(JSON.stringify(mapFile.error.issues));
        return undefined;
    }
    return mapFile.data;
}
export function getMapsScripts(maps) {
    const { scripts } = getMapsScriptsWithAliases(maps);
    return scripts;
}
// Internal function that returns both scripts and the path-to-alias mapping
function getMapsScriptsWithAliases(maps) {
    const scripts = {};
    const scriptPathToAlias = new Map();
    const aliasCount = new Map();
    for (const [mapPath, map] of maps) {
        if (!map.properties) {
            continue;
        }
        const scriptProperty = map.properties.find((property) => property.name === "script");
        if (!scriptProperty || typeof scriptProperty.value !== "string") {
            continue;
        }
        const scriptAbsolutePath = path.resolve(path.dirname(mapPath), scriptProperty.value);
        // Check if this script path has already been registered
        if (scriptPathToAlias.has(scriptAbsolutePath)) {
            // Script is shared, skip adding it again
            continue;
        }
        // Create a unique alias for this script
        const scriptName = path.parse(scriptAbsolutePath).name;
        const scriptDir = path.basename(path.dirname(scriptAbsolutePath));
        const baseAlias = `${scriptDir}-${scriptName}`;
        // Handle alias collisions by appending parent directory names
        let uniqueScriptName = baseAlias;
        let currentPath = path.dirname(scriptAbsolutePath);
        while (aliasCount.has(uniqueScriptName)) {
            const parentDir = path.basename(path.dirname(currentPath));
            uniqueScriptName = `${parentDir}-${uniqueScriptName}`;
            currentPath = path.dirname(currentPath);
        }
        scripts[uniqueScriptName] = scriptAbsolutePath;
        scriptPathToAlias.set(scriptAbsolutePath, uniqueScriptName);
        aliasCount.set(uniqueScriptName, 1);
    }
    return { scripts, scriptPathToAlias };
}
export function getMapsOptimizers(maps, options) {
    const plugins = [];
    const baseDistPath = options?.output?.path ?? "dist";
    const playUrl = options?.playUrl ?? process.env.PLAY_URL ?? "https://play.workadventu.re";
    // Get the script path to alias mapping
    const { scriptPathToAlias } = getMapsScriptsWithAliases(maps);
    for (const [mapPath, map] of maps) {
        const parsedMapPath = path.parse(mapPath);
        const mapName = parsedMapPath.name;
        const mapDirectory = parsedMapPath.dir;
        const distFolder = path.join(baseDistPath, mapDirectory);
        const optionsParsed = {
            logs: 1,
            output: {
                path: distFolder,
                map: {
                    name: mapName,
                },
            },
            ...options,
        };
        if (!optionsParsed.output) {
            optionsParsed.output = {};
        }
        if (!optionsParsed.output.tileset) {
            optionsParsed.output.tileset = {};
        }
        optionsParsed.output.tileset.prefix = `${mapName}-chunk`;
        optionsParsed.output.tileset.suffix = crypto
            .createHash("shake256", { outputLength: 4 })
            .update(Date.now() + mapName)
            .digest("hex");
        plugins.push(mapOptimizer(mapPath, map, distFolder, structuredClone(optionsParsed), baseDistPath, playUrl, scriptPathToAlias));
    }
    return plugins;
}
// Map Optimizer Vite Plugin
function mapOptimizer(mapPath, map, distFolder, optimizeOptions, baseDistPath, playUrl, scriptPathToAlias) {
    let resolvedOutDir;
    let resolvedAssetsDir;
    return {
        name: "map-optimizer",
        apply: "build",
        configResolved(config) {
            resolvedOutDir = path.resolve(config.root, config.build.outDir || baseDistPath);
            resolvedAssetsDir = config.build.assetsDir ?? "assets";
        },
        load() {
            this.addWatchFile(mapPath);
        },
        async writeBundle() {
            await optimize(mapPath, optimizeOptions);
            const mapName = path.parse(mapPath).name;
            const mapExtension = path.parse(mapPath).ext;
            const optimizedMapFilePath = `${distFolder}/${mapName}${mapExtension}`;
            if (!map?.properties) {
                return;
            }
            if (!fs.existsSync(distFolder)) {
                throw new Error(`Cannot find ${distFolder} build folder`);
            }
            if (!fs.existsSync(optimizedMapFilePath)) {
                throw new Error(`Unknown optimized map file on: ${optimizedMapFilePath}`);
            }
            const optimizedMapFile = await fs.promises.readFile(optimizedMapFilePath);
            const optimizedMapObject = JSON.parse(optimizedMapFile.toString());
            const optimizedMapResult = ITiledMap.safeParse(optimizedMapObject);
            if (!optimizedMapResult.success) {
                throw new Error(`Optimized map file is invalid: ${optimizedMapFilePath}`, {
                    cause: optimizedMapResult.error,
                });
            }
            const optimizedMap = optimizedMapResult.data;
            if (!optimizedMap.properties) {
                throw new Error("Undefined properties on map optimized! Something was wrong!");
            }
            const imageProperty = map.properties.find((property) => property.name === "mapImage");
            if (imageProperty && typeof imageProperty.value === "string" && imageProperty.value !== "") {
                const imagePath = path.resolve(path.dirname(mapPath), imageProperty.value);
                if (fs.existsSync(imagePath)) {
                    const newMapImageName = `${mapName}${path.parse(imagePath).ext}`;
                    await fs.promises.copyFile(imagePath, `${distFolder}/${newMapImageName}`);
                    for (const property of optimizedMap.properties) {
                        if (property.name === "mapImage") {
                            property.value = newMapImageName;
                            break;
                        }
                    }
                }
            }
            await fs.promises.mkdir(path.dirname(optimizedMapFilePath), { recursive: true });
            await fs.promises.writeFile(optimizedMapFilePath, JSON.stringify(optimizedMap));
        },
        async closeBundle() {
            const mapName = path.parse(mapPath).name;
            const mapExtension = path.parse(mapPath).ext;
            const optimizedMapFilePath = `${distFolder}/${mapName}${mapExtension}`;
            if (!map?.properties) {
                return;
            }
            if (!fs.existsSync(optimizedMapFilePath)) {
                // If the optimized map is not found, nothing to update.
                return;
            }
            const optimizedMapFile = await fs.promises.readFile(optimizedMapFilePath);
            const optimizedMapObject = JSON.parse(optimizedMapFile.toString());
            const optimizedMapResult = ITiledMap.safeParse(optimizedMapObject);
            if (!optimizedMapResult.success) {
                throw new Error(`Optimized map file is invalid: ${optimizedMapFilePath}`, {
                    cause: optimizedMapResult.error,
                });
            }
            const optimizedMap = optimizedMapResult.data;
            if (!optimizedMap.properties) {
                throw new Error("Undefined properties on map optimized! Something was wrong!");
            }
            const scriptProperty = map.properties.find((property) => property.name === "script");
            if (!scriptProperty || typeof scriptProperty.value !== "string") {
                return;
            }
            const outDir = resolvedOutDir ?? path.resolve(process.cwd(), baseDistPath);
            const assetsDir = resolvedAssetsDir ?? "assets";
            const assetsFolder = path.join(outDir, assetsDir);
            const scriptAbsolutePath = path.resolve(path.dirname(mapPath), scriptProperty.value);
            const uniqueScriptName = scriptPathToAlias.get(scriptAbsolutePath);
            if (!uniqueScriptName) {
                throw new Error(`Cannot find alias for script: ${scriptAbsolutePath}`);
            }
            // Read manifest and/or scan assets
            let compiledJsBasename;
            let manifestPath = path.join(outDir, ".vite", "manifest.json");
            if (!fs.existsSync(manifestPath)) {
                // Fallback for Vite 4 and earlier
                manifestPath = path.join(outDir, "manifest.json");
            }
            if (fs.existsSync(manifestPath)) {
                const manifestRaw = await fs.promises.readFile(manifestPath, "utf-8");
                const manifest = JSON.parse(manifestRaw);
                const manifestEntry = Object.values(manifest).find((entry) => {
                    const file = entry?.file;
                    if (typeof file !== "string")
                        return false;
                    const base = path.basename(file);
                    return base.startsWith(`${uniqueScriptName}-`) && base.endsWith(".js");
                });
                if (manifestEntry?.file) {
                    compiledJsBasename = path.basename(manifestEntry.file);
                }
            }
            if (!compiledJsBasename) {
                const candidate = fs
                    .readdirSync(assetsFolder)
                    .find((asset) => asset.startsWith(`${uniqueScriptName}-`) && asset.endsWith(".js"));
                if (!candidate) {
                    throw new Error(`Undefined ${uniqueScriptName} script file in ${assetsFolder}`);
                }
                compiledJsBasename = candidate;
            }
            // Generate HTML wrapper file alongside the JS, using same base name
            const htmlFileName = compiledJsBasename.replace(/\.js$/i, ".html");
            const htmlFilePath = `${assetsFolder}/${htmlFileName}`;
            const jsRelativePath = `./${compiledJsBasename}`;
            // Basic URL validation
            try {
                new URL(playUrl);
            }
            catch (error) {
                throw new Error(`Invalid playUrl: ${playUrl}`, { cause: error });
            }
            const htmlContent = `<!DOCTYPE html>\n<html>\n  <head>\n    <script src="${playUrl}/iframe_api.js"></script>\n    <script src="${jsRelativePath}"></script>\n  </head>\n  <body>\n\n  </body>\n</html>`;
            await fs.promises.writeFile(htmlFilePath, htmlContent);
            for (const property of optimizedMap.properties) {
                if (property.name === "script") {
                    property.value = path.relative(distFolder, `${assetsFolder}/${htmlFileName}`);
                    break;
                }
            }
            await fs.promises.writeFile(optimizedMapFilePath, JSON.stringify(optimizedMap));
        },
    };
}
