// @workadventure/map-starter-kit-core/dist/server.js
import express from "express";
import * as path from "node:path";
import * as fs from "node:fs";
import cors from "cors";
import { fileURLToPath } from "node:url";
import Mustache from "mustache";
import { exec } from "node:child_process";
import { promisify } from "node:util";
let coreRoot = null;
function getCoreRoot() {
  if (coreRoot !== null) {
    return coreRoot;
  }
  try {
    const dir = path.dirname(fileURLToPath(import.meta.url));
    const candidate = path.dirname(dir);
    const packagePath = path.join(candidate, "package.json");
    if (fs.existsSync(packagePath)) {
      const pkg = JSON.parse(fs.readFileSync(packagePath, "utf-8"));
      if (pkg.name === "@workadventure/map-starter-kit-core") {
        coreRoot = candidate;
        return coreRoot;
      }
    }
  } catch {
  }
  coreRoot = process.cwd();
  return coreRoot;
}
const DEFAULT_META_TITLE = "WorkAdventure Starter Kit";
const DEFAULT_PAGE_TITLE = "WorkAdventure build your map";
const SELF_HOSTED_META_TITLE = "WorkAdventure Starter Kit - Self-hosted";
class FrontController {
  app;
  headScriptsCache = /* @__PURE__ */ new Map();
  constructor(app2) {
    this.app = app2;
    this.setupRoutes();
    this.setupRoutesStep1();
    this.setupRoutesStep2();
    this.setupRoutesStep3();
    this.setupRoutesStep4();
  }
  setupRoutes() {
    this.app.get("/", async (_, res) => {
      try {
        const headScripts = await this.loadHeadScripts("index");
        const isPublishingConfigured = await this.isPublishingConfigured();
        const secretConfig = await this.getSecretConfig();
        res.send(await this.renderTemplate("index", {
          ...this.buildHeadData({ headScripts }),
          isPublishingConfigured,
          mapStorageUrl: secretConfig?.mapStorageUrl ?? ""
        }));
      } catch (error) {
        console.error("Error rendering template:", error);
        res.status(500).send("Error rendering template");
      }
    });
  }
  /**
   * Render a template file
   * @param filename - The filename of the template to render
   * @returns The rendered template
   */
  async renderTemplate(filename, data = {}) {
    const coreRoot2 = getCoreRoot();
    const templatesDir = path.join(coreRoot2, "public/assets/views");
    if (!fs.existsSync(templatesDir)) {
      throw new Error(`Templates directory not found: ${templatesDir}`);
    }
    const templatePath = path.join(templatesDir, `${filename}.html`);
    const template = await fs.promises.readFile(templatePath, "utf-8");
    const partialsDir = path.join(templatesDir, "partials");
    const partials = {};
    if (fs.existsSync(partialsDir)) {
      const partialFiles = await fs.promises.readdir(partialsDir);
      for (const file of partialFiles) {
        if (!file.endsWith(".html")) {
          continue;
        }
        const partialName = path.basename(file, ".html");
        const partialPath = path.join(partialsDir, file);
        partials[partialName] = await fs.promises.readFile(partialPath, "utf-8");
      }
    }
    return Mustache.render(template, data, partials);
  }
  buildHeadData(options = {}) {
    return {
      metaTitle: options.metaTitle ?? DEFAULT_META_TITLE,
      pageTitle: options.pageTitle ?? DEFAULT_PAGE_TITLE,
      headScripts: options.headScripts ?? ""
    };
  }
  async loadHeadScripts(name) {
    if (!name) {
      return "";
    }
    const cached = this.headScriptsCache.get(name);
    if (cached !== void 0) {
      return cached;
    }
    const coreRoot2 = getCoreRoot();
    const scriptPath = path.join(coreRoot2, "public/assets/views/partials/head-scripts", `${name}.html`);
    if (!fs.existsSync(scriptPath)) {
      this.headScriptsCache.set(name, "");
      return "";
    }
    const script = await fs.promises.readFile(scriptPath, "utf-8");
    this.headScriptsCache.set(name, script);
    return script;
  }
  async getSecretConfig() {
    const envSecretPath = path.join(process.cwd(), ".env.secret");
    const hasSecretFile = await fs.promises.access(envSecretPath).then(() => true).catch(() => false);
    if (!hasSecretFile) {
      return null;
    }
    const secretContent = await fs.promises.readFile(envSecretPath, "utf-8");
    const mapStorageUrl = secretContent.match(/MAP_STORAGE_URL=(.+)/)?.[1]?.trim() || null;
    const mapStorageApiKey = secretContent.match(/MAP_STORAGE_API_KEY=(.+)/)?.[1]?.trim() || null;
    const uploadDirectory = secretContent.match(/UPLOAD_DIRECTORY=(.+)/)?.[1]?.trim() || null;
    return {
      mapStorageUrl,
      mapStorageApiKey,
      uploadDirectory
    };
  }
  async isPublishingConfigured() {
    const secretConfig = await this.getSecretConfig();
    return Boolean(
      secretConfig?.mapStorageUrl && secretConfig?.mapStorageApiKey && secretConfig?.uploadDirectory
    );
  }
  /**
   * Setup the routes for file "step1-git.html"
   * @returns void
   */
  setupRoutesStep1() {
    this.app.get("/step1-git", async (_, res) => {
      const headScripts = await this.loadHeadScripts("background-fade");
      res.send(await this.renderTemplate("step1-git", this.buildHeadData({ headScripts })));
    });
  }
  /**
   * Setup the routes for file "step2-hosting.html"
   * @returns void
   */
  setupRoutesStep2() {
    this.app.get("/step2-hosting", async (_, res) => {
      res.send(await this.renderTemplate("step2-hosting", this.buildHeadData()));
    });
  }
  /**
   * Setup the routes for file "step3-steps.html"
   * @returns void
   */
  setupRoutesStep3() {
    this.app.get("/step3-steps", async (_, res) => {
      const headScripts = await this.loadHeadScripts("background-fade");
      const secretConfig = await this.getSecretConfig();
      res.send(await this.renderTemplate("step3-steps", {
        ...this.buildHeadData({ headScripts }),
        mapStorageUrl: secretConfig?.mapStorageUrl ?? "",
        mapStorageApiKey: secretConfig?.mapStorageApiKey ?? "",
        uploadDirectory: secretConfig?.uploadDirectory ?? ""
      }));
    });
    this.app.get("/step3-steps-selfhosted", async (_, res) => {
      const headScripts = await this.loadHeadScripts("background-fade-selfhosted");
      const secretConfig = await this.getSecretConfig();
      res.send(await this.renderTemplate("step3-steps-selfhosted", {
        ...this.buildHeadData({
          metaTitle: SELF_HOSTED_META_TITLE,
          pageTitle: "WorkAdventure test map - Self-hosted",
          headScripts
        }),
        mapStorageUrl: secretConfig?.mapStorageUrl ?? "",
        mapStorageApiKey: secretConfig?.mapStorageApiKey ?? "",
        uploadDirectory: secretConfig?.uploadDirectory ?? ""
      }));
    });
  }
  /**
   * Setup the routes for file "step4-map.html"
   * @returns void
   */
  setupRoutesStep4() {
    this.app.get("/step4-validated", async (_, res) => {
      const headScripts = await this.loadHeadScripts("background-fade");
      res.send(await this.renderTemplate("step4-validated", this.buildHeadData({ headScripts })));
    });
    this.app.get("/step4-validated-selfhosted", async (_, res) => {
      const headScripts = await this.loadHeadScripts("selfhosted-maps");
      res.send(await this.renderTemplate("step4-validated-selfhosted", this.buildHeadData({
        metaTitle: SELF_HOSTED_META_TITLE,
        pageTitle: "Your maps - Self-hosted",
        headScripts
      })));
    });
  }
}
class MapController {
  router;
  app;
  constructor(app2) {
    this.app = app2;
    this.router = express.Router();
    this.setupRoutes();
    this.app.use("/maps", this.router);
  }
  /**
   * Setup the routes for the map controller
   * @returns void
   */
  setupRoutes() {
    this.router.get("/list", async (_, res) => {
      try {
        const mapsDir = "./";
        const maps = await this.getMapsWithProperties(mapsDir);
        res.json(maps);
      } catch (error) {
        console.error("Error getting maps list:", error);
        res.status(500).json({ error: "Error getting maps list" });
      }
    });
  }
  /**
   * Get the list of maps with their properties
   * @param dir - The directory to search for maps
   * @param baseDir - The base directory to use for relative paths
   * @returns The list of maps with their properties
   */
  async getMapsWithProperties(dir, baseDir = dir) {
    let files = await fs.promises.readdir(dir, { withFileTypes: true });
    const maps = [];
    for (const file of files) {
      const fullPath = path.join(dir, file.name);
      if (file.name === "dist") continue;
      if (file.isDirectory()) {
        const subMaps = await this.getMapsWithProperties(fullPath, baseDir);
        maps.push(...subMaps);
      } else if (file.name.endsWith(".tmj")) {
        try {
          const tmjContent = await fs.promises.readFile(fullPath, "utf-8");
          const tmjData = JSON.parse(tmjContent);
          const properties = tmjData.properties || [];
          const findProperty = (key) => {
            const item = properties.find((p) => p.name === key);
            return item ? item.value : null;
          };
          const stats = await fs.promises.stat(fullPath);
          const fileSizeInMB = (stats.size / (1024 * 1024)).toFixed(2);
          const lastModified = stats.mtime;
          const relativePath = path.relative(baseDir, fullPath).replace(/\\/g, "/");
          const filename = path.basename(file.name, ".tmj");
          maps.push({
            path: relativePath,
            filename,
            mapName: findProperty("mapName") || filename,
            mapImage: findProperty("mapImage") || null,
            mapDescription: findProperty("mapDescription") || "",
            mapCopyright: findProperty("mapCopyright") || "",
            size: fileSizeInMB,
            lastModified: lastModified.toISOString(),
            lastModifiedFormatted: lastModified.toLocaleDateString("fr-FR", {
              day: "2-digit",
              month: "2-digit",
              year: "numeric",
              hour: "2-digit",
              minute: "2-digit"
            })
          });
        } catch (error) {
          console.error(`Error reading TMJ file ${fullPath}:`, error);
        }
      }
    }
    return maps;
  }
}
const execAsync = promisify(exec);
function hydrateMapStorageList(raw, mapStorageBaseUrl) {
  const list = [];
  const baseUrl = mapStorageBaseUrl.replace(/\/$/, "");
  for (const [wamFile, entry] of Object.entries(raw)) {
    if (!entry || typeof entry !== "object") continue;
    const rawWamFileUrl = entry.wamFileUrl ?? wamFile;
    const wamFileUrlNormalized = rawWamFileUrl.startsWith("http") ? new URL(rawWamFileUrl).pathname.replace(/^\//, "") : rawWamFileUrl.replace(/^\//, "");
    const filename = wamFileUrlNormalized.split("/").pop() ?? wamFileUrlNormalized;
    const nameFromFile = filename.replace(/\.(tmj|wam)$/i, "");
    const meta = entry.metadata ?? {};
    const mapUrl = entry.mapUrl ?? entry.mapUrl ?? "";
    let mapImage = null;
    if (meta.thumbnail) {
      if (meta.thumbnail.startsWith("http")) {
        mapImage = meta.thumbnail;
      } else {
        const thumbFile = meta.thumbnail.replace(/^\//, "");
        const wamDir = wamFileUrlNormalized.includes("/") ? wamFileUrlNormalized.replace(/\/[^/]*$/, "") : "";
        const thumbPath = wamDir ? `${wamDir}/${thumbFile}` : thumbFile;
        mapImage = `${baseUrl}/${thumbPath}`;
      }
    }
    list.push({
      wamFileUrl: wamFileUrlNormalized,
      filename,
      mapName: meta.name ?? nameFromFile,
      mapDescription: meta.description,
      mapCopyright: meta.copyright,
      mapImage,
      mapUrl: mapUrl.startsWith("http") ? mapUrl : `${baseUrl}/${mapUrl.replace(/^\//, "")}`
    });
  }
  return list;
}
class UploaderController {
  router;
  app;
  constructor(app2) {
    this.app = app2;
    this.router = express.Router();
    this.setupMiddleware();
    this.setupRoutes();
    this.app.use("/uploader", this.router);
  }
  setupMiddleware() {
    this.router.use(express.json());
  }
  setupRoutes() {
    this.router.post("/configure", async (req, res) => {
      try {
        const { mapStorageUrl, mapStorageApiKey, uploadDirectory } = req.body;
        if (!mapStorageUrl || !mapStorageApiKey || !uploadDirectory) {
          return res.status(400).json({
            error: "Missing required fields",
            required: ["mapStorageUrl", "mapStorageApiKey", "uploadDirectory"]
          });
        }
        await this.createEnvSecretFile({
          mapStorageUrl,
          mapStorageApiKey,
          uploadDirectory
        });
        return res.json({
          success: true,
          message: "Configuration updated successfully"
        });
      } catch (error) {
        console.error("Error configuring uploader:", error);
        return res.status(500).json({
          error: "Error configuring uploader",
          message: error instanceof Error ? error.message : "Unknown error"
        });
      }
    });
    this.router.get("/maps-storage-list", async (_, res) => {
      try {
        const envSecretPath = path.join(process.cwd(), ".env.secret");
        const hasSecretFile = await fs.promises.access(envSecretPath).then(() => true).catch(() => false);
        if (!hasSecretFile) {
          return res.status(400).json({
            error: "Configuration not found",
            message: "Please configure the upload settings first."
          });
        }
        const secretContent = await fs.promises.readFile(envSecretPath, "utf-8");
        const mapStorageUrl = secretContent.match(/MAP_STORAGE_URL=(.+)/)?.[1]?.trim();
        const mapStorageApiKey = secretContent.match(/MAP_STORAGE_API_KEY=(.+)/)?.[1]?.trim();
        if (!mapStorageUrl || !mapStorageApiKey) {
          return res.status(400).json({
            error: "Missing map-storage configuration",
            message: "MAP_STORAGE_URL and MAP_STORAGE_API_KEY must be set in .env.secret."
          });
        }
        const baseUrl = mapStorageUrl.replace(/\/$/, "");
        const listUrl = `${baseUrl}/maps/`;
        const response = await fetch(listUrl, {
          method: "GET",
          headers: {
            "Authorization": `Bearer ${mapStorageApiKey}`,
            "Content-Type": "application/json"
          }
        });
        if (!response.ok) {
          const text = await response.text();
          console.error("Map-storage list error:", response.status, text);
          return res.status(response.status).json({
            error: "Map-storage request failed",
            message: response.status === 401 ? "Invalid API key. Check MAP_STORAGE_API_KEY in your .env.secret." : `Map-storage returned ${response.status}: ${text.slice(0, 200)}`
          });
        }
        const data = await response.json();
        const raw = typeof data === "object" && data !== null ? data : {};
        const rawMap = !Array.isArray(raw) && typeof raw.maps === "object" && raw.maps !== null ? raw.maps : !Array.isArray(raw) ? raw : {};
        const maps = hydrateMapStorageList(rawMap, baseUrl);
        const playBaseUrl = secretContent.match(/PLAY_BASE_URL=(.+)/)?.[1]?.trim() || null;
        return res.json({ maps, mapStorageUrl: baseUrl, playBaseUrl });
      } catch (error) {
        console.error("Error fetching maps from map-storage:", error);
        return res.status(500).json({
          error: "Error fetching maps list",
          message: error instanceof Error ? error.message : "Unknown error"
        });
      }
    });
    this.router.post("/upload", async (_, res) => {
      try {
        const envSecretPath = path.join(process.cwd(), ".env.secret");
        const hasSecretFile = await fs.promises.access(envSecretPath).then(() => true).catch(() => false);
        if (!hasSecretFile) {
          return res.status(400).json({
            error: "Configuration not found",
            message: "Please configure the upload settings first using /uploader/configure"
          });
        }
        await this.runUpload();
        return res.json({
          success: true,
          message: "Map uploaded successfully"
        });
      } catch (error) {
        console.error("Error uploading map:", error);
        return res.status(500).json({
          error: "Error uploading map",
          message: error instanceof Error ? error.message : "Unknown error"
        });
      }
    });
  }
  async runUpload() {
    const projectRoot = process.cwd();
    try {
      const { stdout, stderr } = await execAsync("npm run upload", {
        cwd: projectRoot,
        env: {
          ...process.env,
          // Ensure we're using the current process environment
          NODE_ENV: process.env.NODE_ENV || "development"
        },
        maxBuffer: 10 * 1024 * 1024
        // 10MB buffer for output
      });
      if (stderr && !stderr.includes("warning")) {
        console.warn("Upload stderr:", stderr);
      }
      console.info("Upload stdout:", stdout);
      console.info("Upload completed successfully");
    } catch (error) {
      console.error("Error executing upload-only:", error);
      throw error;
    }
  }
  async createEnvSecretFile(config) {
    const envSecretPath = path.join(process.cwd(), ".env.secret");
    const secretContent = `# Secret configuration file for MAP_STORAGE upload mode
# This file is not committed to git (see .gitignore)

MAP_STORAGE_URL=${config.mapStorageUrl}
MAP_STORAGE_API_KEY=${config.mapStorageApiKey}
UPLOAD_DIRECTORY=${config.uploadDirectory}
`;
    await fs.promises.writeFile(envSecretPath, secretContent, "utf-8");
  }
}
const app = express();
const corsOptions = {
  credentials: true,
  // Allow sending cookies
  methods: ["GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS"],
  allowedHeaders: [
    "Content-Type",
    "Authorization",
    "X-Requested-With",
    "Accept",
    "Origin",
    "Access-Control-Request-Method",
    "Access-Control-Request-Headers"
  ],
  exposedHeaders: ["Content-Length", "Content-Type"],
  maxAge: 86400
  // Cache the OPTIONS requests for 24 hours
};
app.use(cors(corsOptions));
app.use(express.json());
const staticOptions = {
  maxAge: "1d",
  // Cache the files for 1 day
  etag: true,
  // Enable ETag for cache validation
  lastModified: true
  // Enable Last-Modified header
};
app.use("/assets", express.static(path.join(process.cwd(), "dist", "assets"), staticOptions));
app.use("/public", express.static(path.join(getCoreRoot(), "public"), staticOptions));
app.use("/tilesets", express.static(path.join(process.cwd(), "tilesets"), {
  maxAge: "7d",
  etag: true,
  lastModified: true
}));
const staticMiddleware = express.static(".", staticOptions);
app.use("/src", async (req, res, next) => {
  if (req.path.endsWith(".ts") || req.path.endsWith(".tsx")) {
    try {
      const filePath = path.join(process.cwd(), "src", req.path.startsWith("/") ? req.path.slice(1) : req.path);
      if (!fs.existsSync(filePath)) {
        return res.status(404).send("File not found");
      }
      const esbuild = await import("esbuild");
      const result = await esbuild.build({
        entryPoints: [filePath],
        bundle: true,
        format: "esm",
        target: "esnext",
        write: false,
        platform: "browser",
        sourcemap: false,
        // Externalize WorkAdventure global API (available in the browser)
        external: ["WA"]
      });
      res.setHeader("Content-Type", "application/javascript; charset=utf-8");
      return res.send(result.outputFiles[0].text);
    } catch (error) {
      console.error("Error transforming TypeScript file:", error);
      return next(error);
    }
  }
  next();
});
app.use((req, res, next) => {
  if (req.path.startsWith("/src/")) {
    return next();
  }
  staticMiddleware(req, res, next);
});
[
  new MapController(app),
  new FrontController(app),
  new UploaderController(app)
];
const viteNodeApp = app;
export {
  app as default,
  viteNodeApp
};
